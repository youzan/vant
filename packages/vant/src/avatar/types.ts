// Types
export type AvatarSize = 'large' | 'medium' | 'normal' | 'small';
export type AvatarShape = 'square' | 'round';

export type AvatarThemeVars = {
  avatarSize?: string;
  avatarSizeLarge?: string;
  avatarSizeMedium?: string;
  avatarSizeSmall?: string;
  avatarBgColor?: string;
  avatarTextColor?: string;
  avatarFontSize?: string;
  avatarFontSizeLarge?: string;
  avatarFontSizeMedium?: string;
  avatarFontSizeSmall?: string;
  avatarFontWeight?: string;
  avatarLineHeight?: string;
  avatarBorderRadius?: string;
};
