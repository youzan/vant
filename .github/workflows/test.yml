name: CI

on:
  push:
    branches:
      - '**'

  pull_request:
    branches:
      - main

  workflow_dispatch:

jobs:
  lint:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v4

      - name: Install pnpm
        run: |
          npm install -g corepack@latest --force 
          corepack enable

      - uses: actions/setup-node@v4
        with:
          node-version: '22.14.0'
          cache: 'pnpm'

      - name: Install dependencies
        run: pnpm install

      - name: Run linter
        run: npm run lint

  test:
    runs-on: macos-14
    steps:
      - uses: actions/checkout@v4

      - name: Install pnpm
        run: |
          npm install -g corepack@latest --force 
          corepack enable

      - uses: actions/setup-node@v4
        with:
          node-version: '22.14.0'
          cache: 'pnpm'

      - name: Install dependencies
        run: pnpm install

      - name: Run test cases
        run: pnpm run test:coverage

      - name: Upload coverage to Codecov
        uses: codecov/codecov-action@v5
        with:
          token: ${{ secrets.CODECOV_TOKEN }}
          directory: ./packages/vant/test/coverage

  build:
    runs-on: macos-14
    steps:
      - uses: actions/checkout@v4

      - name: Install pnpm
        run: |
          npm install -g corepack@latest --force 
          corepack enable

      - uses: actions/setup-node@v4
        with:
          node-version: '22.14.0'
          cache: 'pnpm'

      - name: Install dependencies
        run: pnpm install

      - name: Build
        run: npm run build
