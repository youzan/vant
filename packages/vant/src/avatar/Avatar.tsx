import {
  defineComponent,
  type PropType,
  type CSSProperties,
  type ExtractPropTypes,
} from 'vue';
import { isDef, addUnit, makeStringProp, createNamespace } from '../utils';
import { AvatarShape, AvatarSize } from './types';

const [name, bem] = createNamespace('avatar');

export const avatarProps = {
  src: String,
  text: String,
  size: [Number, String] as PropType<number | string | AvatarSize>,
  shape: makeStringProp<AvatarShape>('round'),
  bgColor: String,
  color: String,
  icon: String,
  alt: String,
};

export type AvatarProps = ExtractPropTypes<typeof avatarProps>;

export default defineComponent({
  name,

  props: avatarProps,

  emits: ['error', 'click'],

  setup(props, { emit, slots }) {
    const style = (): CSSProperties => {
      const style: CSSProperties = {};

      if (isDef(props.size)) {
        const size = addUnit(props.size);
        style.width = size;
        style.height = size;
        style.fontSize = `calc(${size} * 0.45)`;
      }

      if (props.color) {
        style.color = props.color;
      }

      if (props.bgColor) {
        style.backgroundColor = props.bgColor;
        if (!props.color) {
          style.color = '#fff';
        }
      }

      return style;
    };

    const onImageError = (event: Event) => {
      emit('error', event);
    };

    const onImageClick = () => {
      emit('click');
    };

    const renderContent = () => {
      if (slots.default) {
        return slots.default();
      }

      if (props.src) {
        return (
          <img
            src={props.src}
            alt={props.alt}
            class={bem('img')}
            onError={onImageError}
            onClick={onImageClick}
          />
        );
      }

      if (props.text) {
        return <span class={bem('text')}>{props.text}</span>;
      }

      if (props.icon) {
        return <i class={[bem('icon'), `van-icon van-icon-${props.icon}`]} />;
      }

      return null;
    };

    return () => (
      <div class={bem([props.shape, props.size])} style={style()}>
        {renderContent()}
      </div>
    );
  },
});
