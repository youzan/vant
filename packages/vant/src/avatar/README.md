# Avatar

### Intro

Used to represent users or things, supporting image, text, or icon display.

### Install

Register component globally via `app.use`, refer to [Component Registration](#/en-US/advanced-usage#zu-jian-zhu-ce) for more registration ways.

```js
import { createApp } from 'vue';
import { Avatar } from 'vant';

const app = createApp();
app.use(Avatar);
```

## Usage

### Basic Usage

Avatar supports three types: image, text, and icon.

```html
<van-avatar :src="avatarURL" />
<van-avatar text="U" />
<van-avatar icon="user-o" />
```

### Avatar Shape

Use the `shape` prop to set the avatar shape. Supports `round` and `square`, default is `round`.

```html
<van-avatar :src="avatarURL" shape="square" />
<van-avatar :src="avatarURL" shape="round" />
```

### Avatar Size

Use the `size` prop to set the avatar size. Supports preset sizes and custom numeric values.

```html
<!-- Preset sizes -->
<van-avatar :src="avatarURL" size="large" />
<van-avatar :src="avatarURL" size="medium" />
<van-avatar :src="avatarURL" size="normal" />
<van-avatar :src="avatarURL" size="small" />

<!-- Custom sizes -->
<van-avatar :src="avatarURL" size="80" />
<van-avatar :src="avatarURL" :size="60" />
```

### Custom Colors

Use the `bg-color` and `color` props to customize background and text colors.

```html
<van-avatar text="V" bg-color="#DC143C" color="#fff" />
<van-avatar text="A" bg-color="#228B22" color="#fff" />
<van-avatar text="N" bg-color="#1E90FF" color="#fff" />
<van-avatar text="T" bg-color="#8A2BE2" color="#fff" />
<van-avatar icon="user-o" bg-color="#EEEEEE" color="#7B8198" />
```

### Avatar with Badge

Use with Badge component to display badges on the top-right corner of the avatar.

```html
<van-badge content="10" position="top-left">
  <van-avatar :src="avatarURL" shape="square" />
</van-badge>

<van-badge :content="20">
  <van-avatar :src="avatarURL" shape="square" />
</van-badge>

<van-badge dot>
  <van-avatar :src="avatarURL" shape="square" />
</van-badge>

<van-badge dot>
  <van-avatar text="A" shape="square" />
</van-badge>
```

### Avatar Group

Use AvatarGroup component to display multiple avatars with cascading effects.

```html
<!-- Basic Avatar Group -->
<van-avatar-group :max-count="4">
  <van-avatar :src="avatarURL" />
  <van-avatar :src="avatarURL2" />
  <van-avatar text="A" bg-color="#1E90FF" />
  <van-avatar text="B" bg-color="#228B22" />
  <van-avatar text="C" bg-color="#DC143C" />
  <van-avatar icon="user-o" bg-color="#8A2BE2" />
</van-avatar-group>

<!-- Left-up Cascading -->
<van-avatar-group :max-count="4" cascading="left-up">
  <van-avatar :src="avatarURL" />
  <van-avatar :src="avatarURL2" />
  <van-avatar text="A" bg-color="#1E90FF" />
  <van-avatar text="B" bg-color="#228B22" />
  <van-avatar text="C" bg-color="#DC143C" />
  <van-avatar icon="user-o" bg-color="#8A2BE2" />
</van-avatar-group>

<!-- Right-up Cascading -->
<van-avatar-group :max-count="4" cascading="right-up">
  <van-avatar :src="avatarURL" />
  <van-avatar :src="avatarURL2" />
  <van-avatar text="A" bg-color="#1E90FF" />
  <van-avatar text="B" bg-color="#228B22" />
  <van-avatar text="C" bg-color="#DC143C" />
  <van-avatar icon="user-o" bg-color="#8A2BE2" />
</van-avatar-group>
```

### Group Size and Shape

Use AvatarGroup's `size` and `shape` props to set the size and shape of avatars in the group uniformly. Individual avatar props have higher priority.

```html
<!-- Group Size -->
<van-avatar-group :max-count="4" size="small">
  <van-avatar :src="avatarURL" />
  <van-avatar :src="avatarURL2" />
  <van-avatar text="A" bg-color="#1E90FF" />
  <van-avatar text="B" bg-color="#228B22" />
  <van-avatar text="C" bg-color="#DC143C" />
</van-avatar-group>

<!-- Group Shape -->
<van-avatar-group :max-count="4" shape="square" collapseAvatar="Custom">
  <van-avatar :src="avatarURL" shape="round" />
  <van-avatar :src="avatarURL2" />
  <van-avatar text="A" bg-color="#1E90FF" shape="round" />
  <van-avatar text="B" bg-color="#228B22" />
  <van-avatar text="C" bg-color="#DC143C" />
</van-avatar-group>
```

### Custom Collapse Avatar

When the number of avatars exceeds the limit, you can customize the collapsed avatar display through the `collapse-avatar` slot.

```html
<van-avatar-group :max-count="4" cascading="right-up" size="45">
  <van-avatar :src="avatarURL" />
  <van-avatar :src="avatarURL2" />
  <van-avatar text="A" bg-color="#1E90FF" />
  <van-avatar text="B" bg-color="#228B22" />
  <van-avatar text="C" bg-color="#DC143C" />
  <van-avatar icon="user-o" bg-color="#8A2BE2" />
  <template #collapseAvatar="{ restCount }">
    <van-avatar :text="`+${restCount}`" bg-color="#CCCCCC" />
  </template>
</van-avatar-group>
```

## API

### Avatar Props

| Attribute | Description | Type | Default |
| --- | --- | --- | --- |
| src | Avatar image URL | _string_ | - |
| text | Avatar text content | _string_ | - |
| icon | Avatar icon name | _string_ | - |
| shape | Avatar shape, can be set to `round` `square` | _string_ | `round` |
| size | Avatar size, can be set to `large` `medium` `normal` `small`, also supports numeric values | _number \| string_ | `normal` |
| bg-color | Background color | _string_ | - |
| color | Text color | _string_ | `#fff` |
| alt | Alternative text for failed image loading | _string_ | - |

### AvatarGroup Props

| Attribute | Description | Type | Default |
| --- | --- | --- | --- |
| max-count | Maximum number of avatars to display | _number \| string_ | - |
| cascading | Cascading relationship between images, can be set to `left-up` `right-up` | _string_ | `left-up` |
| shape | Shape, lower priority than Avatar.shape | _string_ | - |
| size | Size, lower priority than Avatar.size | _number \| string_ | - |
| collapse-avatar | Text content for collapsed element when avatar count exceeds limit | _string_ | - |

### Avatar Events

| Event | Description                      | Arguments      |
| ----- | -------------------------------- | -------------- |
| error | Emitted when image fails to load | _event: Event_ |
| click | Emitted when avatar is clicked   | -              |

### Avatar Slots

| Name    | Description           |
| ------- | --------------------- |
| default | Custom avatar content |

### AvatarGroup Slots

| Name            | Description                     | SlotProps               |
| --------------- | ------------------------------- | ----------------------- |
| default         | Avatar group content            | -                       |
| collapse-avatar | Custom collapsed avatar content | _{ restCount: number }_ |

### Types

The component exports the following type definitions:

```ts
import type {
  AvatarProps,
  AvatarSize,
  AvatarShape,
  AvatarGroupProps,
} from 'vant';
```

## Theming

### CSS Variables

The component provides the following CSS variables, which can be used to customize styles. Please refer to [ConfigProvider component](#/en-US/config-provider).

#### Avatar CSS Variables

| Name | Default Value | Description |
| --- | --- | --- |
| --van-avatar-size | _54px_ | Normal size |
| --van-avatar-size-large | _76px_ | Large size |
| --van-avatar-size-medium | _64px_ | Medium size |
| --van-avatar-size-small | _48px_ | Small size |
| --van-avatar-bg-color | _var(--van-gray-5)_ | Default background color |
| --van-avatar-text-color | _var(--van-white)_ | Default text color |
| --van-avatar-font-size | _20px_ | Normal font size |
| --van-avatar-font-size-large | _28px_ | Large font size |
| --van-avatar-font-size-medium | _24px_ | Medium font size |
| --van-avatar-font-size-small | _18px_ | Small font size |
| --van-avatar-font-weight | _var(--van-font-bold)_ | Font weight |
| --van-avatar-line-height | _1.2_ | Line height |
| --van-avatar-border-radius | _var(--van-radius-md)_ | Avatar border radius |

#### AvatarGroup CSS Variables

| Name                       | Default Value | Description             |
| -------------------------- | ------------- | ----------------------- |
| --van-avatar-group-overlap | _-12px_       | Avatar overlap distance |
