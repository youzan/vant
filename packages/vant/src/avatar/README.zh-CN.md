# Avatar 头像

### 介绍

用来代表用户或事物，支持图片、文本或图标展示。

### 引入

通过以下方式来全局注册组件，更多注册方式请参考[组件注册](#/zh-CN/advanced-usage#zu-jian-zhu-ce)。

```js
import { createApp } from 'vue';
import { Avatar } from 'vant';

const app = createApp();
app.use(Avatar);
```

## 代码演示

### 基础用法

头像支持三种类型：图片、文本和图标。

```html
<van-avatar :src="avatarURL" />
<van-avatar text="U" />
<van-avatar icon="user-o" />
```

### 头像形状

通过 `shape` 属性可以设置头像形状，支持 `round` 圆形和 `square` 方形，默认为圆形。

```html
<van-avatar :src="avatarURL" shape="square" />
<van-avatar :src="avatarURL" shape="round" />
```

### 头像尺寸

通过 `size` 属性可以设置头像尺寸，支持预设尺寸和自定义数值。

```html
<!-- 预设尺寸 -->
<van-avatar :src="avatarURL" size="large" />
<van-avatar :src="avatarURL" size="medium" />
<van-avatar :src="avatarURL" size="normal" />
<van-avatar :src="avatarURL" size="small" />

<!-- 自定义尺寸 -->
<van-avatar :src="avatarURL" size="80" />
<van-avatar :src="avatarURL" :size="60" />
```

### 自定义颜色

通过 `bg-color` 和 `color` 属性可以自定义背景色和文字颜色。

```html
<van-avatar text="V" bg-color="#DC143C" color="#fff" />
<van-avatar text="A" bg-color="#228B22" color="#fff" />
<van-avatar text="N" bg-color="#1E90FF" color="#fff" />
<van-avatar text="T" bg-color="#8A2BE2" color="#fff" />
<van-avatar icon="user-o" bg-color="#EEEEEE" color="#7B8198" />
```

### 带徽记的头像

结合 Badge 组件使用，可以在头像右上角显示徽记。

```html
<van-badge content="10" position="top-left">
  <van-avatar :src="avatarURL" shape="square" />
</van-badge>

<van-badge :content="20">
  <van-avatar :src="avatarURL" shape="square" />
</van-badge>

<van-badge dot>
  <van-avatar :src="avatarURL" shape="square" />
</van-badge>

<van-badge dot>
  <van-avatar text="A" shape="square" />
</van-badge>
```

### 头像组

使用 AvatarGroup 组件可以将多个头像组合展示，支持层叠效果。

```html
<!-- 基础头像组 -->
<van-avatar-group :max-count="4">
  <van-avatar :src="avatarURL" />
  <van-avatar :src="avatarURL2" />
  <van-avatar text="A" bg-color="#1E90FF" />
  <van-avatar text="B" bg-color="#228B22" />
  <van-avatar text="C" bg-color="#DC143C" />
  <van-avatar icon="user-o" bg-color="#8A2BE2" />
</van-avatar-group>

<!-- 左侧叠层 -->
<van-avatar-group :max-count="4" cascading="left-up">
  <van-avatar :src="avatarURL" />
  <van-avatar :src="avatarURL2" />
  <van-avatar text="A" bg-color="#1E90FF" />
  <van-avatar text="B" bg-color="#228B22" />
  <van-avatar text="C" bg-color="#DC143C" />
  <van-avatar icon="user-o" bg-color="#8A2BE2" />
</van-avatar-group>

<!-- 右侧叠层 -->
<van-avatar-group :max-count="4" cascading="right-up">
  <van-avatar :src="avatarURL" />
  <van-avatar :src="avatarURL2" />
  <van-avatar text="A" bg-color="#1E90FF" />
  <van-avatar text="B" bg-color="#228B22" />
  <van-avatar text="C" bg-color="#DC143C" />
  <van-avatar icon="user-o" bg-color="#8A2BE2" />
</van-avatar-group>
```

### 组尺寸和组形状

通过 AvatarGroup 的 `size` 和 `shape` 属性可以统一设置组内头像的尺寸和形状，子头像的属性优先级更高。

```html
<!-- 组尺寸 -->
<van-avatar-group :max-count="4" size="small">
  <van-avatar :src="avatarURL" />
  <van-avatar :src="avatarURL2" />
  <van-avatar text="A" bg-color="#1E90FF" />
  <van-avatar text="B" bg-color="#228B22" />
  <van-avatar text="C" bg-color="#DC143C" />
</van-avatar-group>

<!-- 组形状 -->
<van-avatar-group :max-count="4" shape="square" collapseAvatar="自定义">
  <van-avatar :src="avatarURL" shape="round" />
  <van-avatar :src="avatarURL2" />
  <van-avatar text="A" bg-color="#1E90FF" shape="round" />
  <van-avatar text="B" bg-color="#228B22" />
  <van-avatar text="C" bg-color="#DC143C" />
</van-avatar-group>
```

### 自定义折叠头像

当头像数量超出限制时，可以通过 `collapse-avatar` 插槽自定义折叠头像的展示。

```html
<van-avatar-group :max-count="4" cascading="right-up" size="45">
  <van-avatar :src="avatarURL" />
  <van-avatar :src="avatarURL2" />
  <van-avatar text="A" bg-color="#1E90FF" />
  <van-avatar text="B" bg-color="#228B22" />
  <van-avatar text="C" bg-color="#DC143C" />
  <van-avatar icon="user-o" bg-color="#8A2BE2" />
  <template #collapseAvatar="{ restCount }">
    <van-avatar :text="`+${restCount}`" bg-color="#CCCCCC" />
  </template>
</van-avatar-group>
```

## API

### Avatar Props

| 参数 | 说明 | 类型 | 默认值 |
| --- | --- | --- | --- |
| src | 头像图片链接 | _string_ | - |
| text | 头像文字内容 | _string_ | - |
| icon | 头像图标名称 | _string_ | - |
| shape | 头像形状，可选值为 `round` `square` | _string_ | `round` |
| size | 头像尺寸，可选值为 `large` `medium` `normal` `small`，也支持数值 | _number \| string_ | `normal` |
| bg-color | 背景颜色 | _string_ | - |
| color | 文字颜色 | _string_ | `#fff` |
| alt | 图片加载失败替代文本 | _string_ | - |

### AvatarGroup Props

| 参数 | 说明 | 类型 | 默认值 |
| --- | --- | --- | --- |
| max-count | 显示的最大头像个数 | _number \| string_ | - |
| cascading | 图片之间的层叠关系，可选值为 `left-up` `right-up` | _string_ | `left-up` |
| shape | 形状，优先级低于 Avatar.shape | _string_ | - |
| size | 尺寸，优先级低于 Avatar.size | _number \| string_ | - |
| collapse-avatar | 头像数量超出时的折叠元素内容文本 | _string_ | - |

### Avatar Events

| 事件名 | 说明               | 回调参数       |
| ------ | ------------------ | -------------- |
| error  | 图片加载失败时触发 | _event: Event_ |
| click  | 点击头像时触发     | -              |

### Avatar Slots

| 名称    | 描述           |
| ------- | -------------- |
| default | 自定义头像内容 |

### AvatarGroup Slots

| 名称            | 描述               | 参数                    |
| --------------- | ------------------ | ----------------------- |
| default         | 头像组内容         | -                       |
| collapse-avatar | 自定义折叠头像内容 | _{ restCount: number }_ |

### 类型定义

组件导出以下类型定义：

```ts
import type {
  AvatarProps,
  AvatarSize,
  AvatarShape,
  AvatarGroupProps,
} from 'vant';
```

## 主题定制

### 样式变量

组件提供了下列 CSS 变量，可用于自定义样式，使用方法请参考 [ConfigProvider 组件](#/zh-CN/config-provider)。

#### Avatar 样式变量

| 名称                          | 默认值                 | 描述         |
| ----------------------------- | ---------------------- | ------------ |
| --van-avatar-size             | _54px_                 | 普通尺寸     |
| --van-avatar-size-large       | _76px_                 | 大号尺寸     |
| --van-avatar-size-medium      | _64px_                 | 中号尺寸     |
| --van-avatar-size-small       | _48px_                 | 小号尺寸     |
| --van-avatar-bg-color         | _var(--van-gray-5)_    | 默认背景色   |
| --van-avatar-text-color       | _var(--van-white)_     | 默认文字颜色 |
| --van-avatar-font-size        | _20px_                 | 普通字体大小 |
| --van-avatar-font-size-large  | _28px_                 | 大号字体大小 |
| --van-avatar-font-size-medium | _24px_                 | 中号字体大小 |
| --van-avatar-font-size-small  | _18px_                 | 小号字体大小 |
| --van-avatar-font-weight      | _var(--van-font-bold)_ | 文字粗细     |
| --van-avatar-line-height      | _1.2_                  | 行高         |
| --van-avatar-border-radius    | _var(--van-radius-md)_ | 头像圆角     |

#### AvatarGroup 样式变量

| 名称                       | 默认值  | 描述         |
| -------------------------- | ------- | ------------ |
| --van-avatar-group-overlap | _-12px_ | 头像重叠距离 |
