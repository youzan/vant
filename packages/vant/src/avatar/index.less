:root,
:host {
  --van-avatar-size: 54px;
  --van-avatar-size-large: 76px;
  --van-avatar-size-medium: 64px;
  --van-avatar-size-small: 48px;
  --van-avatar-bg-color: var(--van-gray-5);
  --van-avatar-text-color: var(--van-white);
  --van-avatar-font-size: 20px;
  --van-avatar-font-size-large: 28px;
  --van-avatar-font-size-medium: 24px;
  --van-avatar-font-size-small: 18px;
  --van-avatar-font-weight: var(--van-font-bold);
  --van-avatar-line-height: 1.2;
  --van-avatar-border-radius: var(--van-radius-md);
}

.van-avatar {
  display: inline-flex;
  align-items: center;
  justify-content: center;
  width: var(--van-avatar-size);
  height: var(--van-avatar-size);
  color: var(--van-avatar-text-color);
  font-weight: var(--van-avatar-font-weight);
  font-size: var(--van-avatar-font-size);
  line-height: var(--van-avatar-line-height);
  text-align: center;
  background-color: var(--van-avatar-bg-color);
  border-radius: var(--van-avatar-border-radius);
  overflow: hidden;
  user-select: none;
  vertical-align: middle;

  &--round {
    border-radius: 50%;
  }

  &--square {
    border-radius: var(--van-avatar-border-radius);
  }

  &--large {
    width: var(--van-avatar-size-large);
    height: var(--van-avatar-size-large);
    font-size: var(--van-avatar-font-size-large);
  }

  &--medium {
    width: var(--van-avatar-size-medium);
    height: var(--van-avatar-size-medium);
    font-size: var(--van-avatar-font-size-medium);
  }

  &--small {
    width: var(--van-avatar-size-small);
    height: var(--van-avatar-size-small);
    font-size: var(--van-avatar-font-size-small);
  }

  &__img {
    width: 100%;
    height: 100%;
    object-fit: cover;
    display: block;
  }

  &__text {
    white-space: nowrap;
  }

  &__icon {
    font-size: inherit;
  }
}
