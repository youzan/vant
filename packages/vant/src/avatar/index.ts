import { withInstall } from '../utils';
import _Avatar from './Avatar';

export const Avatar = withInstall(_Avatar);
export default Avatar;
export { avatarProps } from './Avatar';
export type { AvatarProps } from './Avatar';
export type { AvatarSize, AvatarShape, AvatarThemeVars } from './types';

declare module 'vue' {
  export interface GlobalComponents {
    VanAvatar: typeof Avatar;
  }
}
