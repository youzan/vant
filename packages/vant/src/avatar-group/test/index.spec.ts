import { h } from 'vue';
import { AvatarGroup } from '..';
import { Avatar } from '../../avatar';
import { mount } from '../../../test';

test('should render avatar group correctly', () => {
  const wrapper = mount(AvatarGroup, {
    slots: {
      default: () => [
        h(Avatar, { src: 'avatar1.jpg' }),
        h(Avatar, { src: 'avatar2.jpg' }),
        h(Avatar, { src: 'avatar3.jpg' }),
      ],
    },
  });

  expect(wrapper.html()).toMatchSnapshot();
  expect(wrapper.findAll('.van-avatar')).toHaveLength(3);
});

test('should limit avatars when maxCount is set', () => {
  const wrapper = mount(AvatarGroup, {
    props: {
      maxCount: 2,
    },
    slots: {
      default: () => [
        h(Avatar, { src: 'avatar1.jpg' }),
        h(Avatar, { src: 'avatar2.jpg' }),
        h(Avatar, { src: 'avatar3.jpg' }),
        h(Avatar, { src: 'avatar4.jpg' }),
      ],
    },
  });

  expect(wrapper.html()).toMatchSnapshot();

  // Should show 2 avatars + 1 collapse avatar
  const avatars = wrapper.findAll('.van-avatar');
  expect(avatars).toHaveLength(3);

  // The last avatar should show the count
  expect(wrapper.text()).toContain('+2');
});

test('should not show collapse avatar when total equals maxCount', () => {
  const wrapper = mount(AvatarGroup, {
    props: {
      maxCount: 3,
    },
    slots: {
      default: () => [
        h(Avatar, { src: 'avatar1.jpg' }),
        h(Avatar, { src: 'avatar2.jpg' }),
        h(Avatar, { src: 'avatar3.jpg' }),
      ],
    },
  });

  const avatars = wrapper.findAll('.van-avatar');
  expect(avatars).toHaveLength(3);
  expect(wrapper.text()).not.toContain('+');
});

test('should apply cascading prop correctly', () => {
  const wrapper = mount(AvatarGroup, {
    props: {
      cascading: 'right-up',
    },
    slots: {
      default: () => [
        h(Avatar, { src: 'avatar1.jpg' }),
        h(Avatar, { src: 'avatar2.jpg' }),
      ],
    },
  });

  expect(wrapper.classes()).toContain('van-avatar-group__right-up');

  // Test left-up cascading
  const wrapper2 = mount(AvatarGroup, {
    props: {
      cascading: 'left-up',
    },
    slots: {
      default: () => [
        h(Avatar, { src: 'avatar1.jpg' }),
        h(Avatar, { src: 'avatar2.jpg' }),
      ],
    },
  });

  expect(wrapper2.classes()).toContain('van-avatar-group__left-up');
});

test('should pass size prop to child avatars', () => {
  const wrapper = mount(AvatarGroup, {
    props: {
      size: 'large',
    },
    slots: {
      default: () => [
        h(Avatar, { src: 'avatar1.jpg' }),
        h(Avatar, { src: 'avatar2.jpg' }),
      ],
    },
  });

  const avatars = wrapper.findAll('.van-avatar');
  avatars.forEach((avatar) => {
    expect(avatar.classes()).toContain('van-avatar--large');
  });
});

test('should pass shape prop to child avatars', () => {
  const wrapper = mount(AvatarGroup, {
    props: {
      shape: 'square',
    },
    slots: {
      default: () => [
        h(Avatar, { src: 'avatar1.jpg' }),
        h(Avatar, { src: 'avatar2.jpg' }),
      ],
    },
  });

  const avatars = wrapper.findAll('.van-avatar');
  avatars.forEach((avatar) => {
    expect(avatar.classes()).toContain('van-avatar--square');
  });
});

test('should pass numeric size prop to child avatars', () => {
  const wrapper = mount(AvatarGroup, {
    props: {
      size: 60,
    },
    slots: {
      default: () => [
        h(Avatar, { src: 'avatar1.jpg' }),
        h(Avatar, { src: 'avatar2.jpg' }),
      ],
    },
  });

  const avatars = wrapper.findAll('.van-avatar');
  avatars.forEach((avatar) => {
    expect(avatar.element.style.width).toBe('60px');
    expect(avatar.element.style.height).toBe('60px');
  });
});

test('should prioritize child avatar props over group props', () => {
  const wrapper = mount(AvatarGroup, {
    props: {
      size: 'large',
      shape: 'square',
    },
    slots: {
      default: () => [
        h(Avatar, { src: 'avatar1.jpg', size: 'small', shape: 'round' }),
        h(Avatar, { src: 'avatar2.jpg' }),
      ],
    },
  });

  const avatars = wrapper.findAll('.van-avatar');

  // First avatar should keep its own props (small, round)
  expect(avatars[0].classes()).toContain('van-avatar--small');
  expect(avatars[0].classes()).not.toContain('van-avatar--square');
  expect(avatars[0].classes()).not.toContain('van-avatar--large');

  // Second avatar should use group props (large, square)
  expect(avatars[1].classes()).toContain('van-avatar--large');
  expect(avatars[1].classes()).toContain('van-avatar--square');
});

test('should render empty group when no children provided', () => {
  const wrapper = mount(AvatarGroup);

  expect(wrapper.html()).toMatchSnapshot();
  expect(wrapper.findAll('.van-avatar')).toHaveLength(0);
});

test('should handle single avatar correctly', () => {
  const wrapper = mount(AvatarGroup, {
    props: {
      maxCount: 5,
    },
    slots: {
      default: () => [h(Avatar, { src: 'avatar1.jpg' })],
    },
  });

  expect(wrapper.findAll('.van-avatar')).toHaveLength(1);
  expect(wrapper.text()).not.toContain('+');
});

test('should handle non-avatar children gracefully', () => {
  const wrapper = mount(AvatarGroup, {
    slots: {
      default: () => [
        h(Avatar, { src: 'avatar1.jpg' }),
        h('div', 'Not an avatar'),
        h(Avatar, { src: 'avatar2.jpg' }),
      ],
    },
  });

  // Should still render all children
  expect(wrapper.findAll('.van-avatar')).toHaveLength(2);
  expect(wrapper.text()).toContain('Not an avatar');
});

test('should calculate collapse count correctly with non-avatar children', () => {
  const wrapper = mount(AvatarGroup, {
    props: {
      maxCount: 2,
    },
    slots: {
      default: () => [
        h(Avatar, { src: 'avatar1.jpg' }),
        h('div', 'Not an avatar'),
        h(Avatar, { src: 'avatar2.jpg' }),
        h(Avatar, { src: 'avatar3.jpg' }),
        h(Avatar, { src: 'avatar4.jpg' }),
      ],
    },
  });

  // Should show 2 children (1 avatar + 1 div) + collapse avatar showing +3 (remaining children)
  const avatars = wrapper.findAll('.van-avatar');
  expect(avatars).toHaveLength(2); // 1 displayed avatar + 1 collapse avatar
  expect(wrapper.text()).toContain('+3');
});

test('should apply z-index correctly for cascading effect', () => {
  const wrapper = mount(AvatarGroup, {
    props: {
      cascading: 'right-up',
    },
    slots: {
      default: () => [
        h(Avatar, { src: 'avatar1.jpg' }),
        h(Avatar, { src: 'avatar2.jpg' }),
        h(Avatar, { src: 'avatar3.jpg' }),
      ],
    },
  });

  const items = wrapper.findAll('.van-avatar-group__item');

  // First item should have highest z-index for right-up
  expect((items[0].element as HTMLElement).style.zIndex).toBe('3');
  expect((items[1].element as HTMLElement).style.zIndex).toBe('2');
  expect((items[2].element as HTMLElement).style.zIndex).toBe('1');
});

test('should handle maxCount of 0', () => {
  const wrapper = mount(AvatarGroup, {
    props: {
      maxCount: 0,
    },
    slots: {
      default: () => [
        h(Avatar, { src: 'avatar1.jpg' }),
        h(Avatar, { src: 'avatar2.jpg' }),
      ],
    },
  });

  // Should show all avatars when maxCount is 0 (same as negative values)
  const avatars = wrapper.findAll('.van-avatar');
  expect(avatars).toHaveLength(2);
  expect(wrapper.text()).not.toContain('+');
});

test('should handle negative maxCount', () => {
  const wrapper = mount(AvatarGroup, {
    props: {
      maxCount: -1,
    },
    slots: {
      default: () => [
        h(Avatar, { src: 'avatar1.jpg' }),
        h(Avatar, { src: 'avatar2.jpg' }),
      ],
    },
  });

  // Should show all avatars when maxCount is negative
  expect(wrapper.findAll('.van-avatar')).toHaveLength(2);
  expect(wrapper.text()).not.toContain('+');
});
