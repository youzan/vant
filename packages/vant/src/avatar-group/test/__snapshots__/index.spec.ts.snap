// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should limit avatars when maxCount is set 1`] = `
<div class="van-avatar-group van-avatar-group__left-up">
  <div
    class="van-avatar-group__item"
    style="z-index: 1;"
  >
    <div class="van-avatar van-avatar--round">
      <img
        src="avatar1.jpg"
        class="van-avatar__img"
      >
    </div>
  </div>
  <div
    class="van-avatar-group__item"
    style="z-index: 2;"
  >
    <div class="van-avatar van-avatar--round">
      <img
        src="avatar2.jpg"
        class="van-avatar__img"
      >
    </div>
  </div>
  <div
    class="van-avatar-group__item"
    style="z-index: 3;"
  >
    <div class="van-avatar van-avatar--round van-avatar-group__collapse">
      <span class="van-avatar__text">
        +2
      </span>
    </div>
  </div>
</div>
`;

exports[`should render avatar group correctly 1`] = `
<div class="van-avatar-group van-avatar-group__left-up">
  <div
    class="van-avatar-group__item"
    style="z-index: 1;"
  >
    <div class="van-avatar van-avatar--round">
      <img
        src="avatar1.jpg"
        class="van-avatar__img"
      >
    </div>
  </div>
  <div
    class="van-avatar-group__item"
    style="z-index: 2;"
  >
    <div class="van-avatar van-avatar--round">
      <img
        src="avatar2.jpg"
        class="van-avatar__img"
      >
    </div>
  </div>
  <div
    class="van-avatar-group__item"
    style="z-index: 3;"
  >
    <div class="van-avatar van-avatar--round">
      <img
        src="avatar3.jpg"
        class="van-avatar__img"
      >
    </div>
  </div>
</div>
`;

exports[`should render empty group when no children provided 1`] = `
<div class="van-avatar-group van-avatar-group__left-up">
</div>
`;
